rainfall_runoff_IHACRES.R

```diff
@@ -0,0 +1,241 @@
+test_data <- window(dat, start = "2011-01-01", end = "2023-12-31")
+
+
+################################################################################
+# Define some objective functions for parameter fitting
+
+# sum of squared errors (SSE)
+SSE <- function(estimated, observed) {
+  sum((estimated - observed)^2,
+      na.rm=TRUE)
+}
+
+# RMSE
+RMSE <- function(estimated, observed) {
+  squared_diff <- (estimated - observed)^2
+  return(sqrt(mean(squared_diff)))
+}
+
+MAE <- function(estimated,observed) {
+  mean(abs(estimated - observed))
+}
+
+
+NSE_R2 <- function(estimated,observed) {
+#r.squared R Squared (Nash-Sutcliffe Efficiency), 
+# 1−∑(Q-X)^2/∑(Q-Q_mean)^2
+# better for R2
+  ref <- mean(observed)
+  err <- abs(observed - estimated)^2
+  referr <- abs(observed - ref)^2
+  1 - sum(err)/sum(referr)
+}
+
+
+BIAS <- function(estimated, observed) {
+# good for matching mass balance -- however,
+# in this case, MAE is better because error is + and -
+  mean(estimated - observed)
+}
+
+
+R2_LOG <- function(estimated,observed) {
+# R2 with log transform of data + offset where
+# offset is Xth percentile of data - here 0.1 or 0.01
+  offset <- quantile(observed,.1)
+  log_observed <- log(observed + offset)
+  log_estimated <- log(estimated + offset)
+  lm_fit <- lm(log_observed ~ log_estimated)
+  r_squared <- summary(lm_fit)$r.squared
+  return(r_squared)
+}
+
+
+################################################################################
+# Run optimization
+
+# Some optimization references
+#https://www.openanalytics.eu/blog/2022/05/17/new-nonlinear-ls-solvers-in-r/
+
+#' method 1: 
+#' stats::optim - General-purpose optimization based on Nelder–Mead, 
+#' quasi-Newton and conjugate-gradient algorithms. It includes an option for 
+#' box-constrained optimization and simulated annealing.
+
+# tau_q, tau_s, vs, delay, loss
+optim_par <- optim(fn = fit_fun,
+           method = 'L-BFGS-B',
+           par = c(3, 50, 0.5, 24, 0.5), 
+           lower = c(0.1, 50, 0.01, 1, 0.01),
+           upper = c(20, 120, .8, 50, 1),
+           control = c("trace" = 6, "factr" = .001),
+           obj = "MAE")# function
+# Plot results
+print(optim_par$par)
+ggarrange(fit_fun(params = optim_par$par, plot = TRUE, return_dat = FALSE),
+          fit_fun(params = optim_par$par, plot = TRUE, return_dat = FALSE, 
+                  dat = "test_data"), nrow = 2)
+
+
+#' Method 2:
+#' stats::nlminb - Unconstrained and box-constrained optimization using PORT 
+#' routines.
+nlminb_par <- nlminb(start = c(3, 50, 0.5, 24, 0.5), 
+            fit_fun,
+            lower = c(0.1, 50, 0.01, 1, 0.01),
+            upper = c(20, 120, .8, 50, 1),
+            control = list("trace" = 6, "iter.max" = 200))
+
+print(nlminb_par$par)
+ggarrange(fit_fun(params = nlminb_par$par, plot = TRUE, return_dat = FALSE),
+          fit_fun(params = nlminb_par$par, plot = TRUE, return_dat = FALSE, 
+                  dat = "test_data"), nrow = 2)
+
+#' Method 3: 
+#' Manual
```

Comment:
Wow, so much better than the ones before it.... how were these values selected by you? Could there be a way to relate these values to the geology/some other aspect of the system?

Reply:
Trying to find an objective function that matches what we know to be important.  Open to suggestions here and actively figuring this out.